doc/port_auto_neg/port-auto-negotiation-design.md

```diff
@@ -0,0 +1,500 @@
+     */
+    SAI_PORT_ATTR_AUTO_NEG_MODE,
+    /**
+     * @brief Speed in Mbps
+     *
+     * On get, returns the configured port speed.
+     *
+     * @type sai_uint32_t
+     * @flags MANDATORY_ON_CREATE | CREATE_AND_SET
+     */
+    SAI_PORT_ATTR_SPEED,
+    /**
+     * @brief Query/Configure list of Advertised port speed (Full-Duplex) in Mbps
+     *
+     * Used when auto negotiation is on. Empty list means all supported values are enabled.
+     *
+     * @type sai_u32_list_t
+     * @flags CREATE_AND_SET
+     * @default empty
+     */
+    SAI_PORT_ATTR_ADVERTISED_SPEED,
+    /**
+     * @brief Configure Interface type
+     *
+     * @type sai_port_interface_type_t
+     * @flags CREATE_AND_SET
+     * @default SAI_PORT_INTERFACE_TYPE_NONE
+     */
+    SAI_PORT_ATTR_INTERFACE_TYPE,
+
+    /**
+     * @brief Configure advertised interface type list
+     *
+     * Used when auto negotiation is on. Empty list means all supported values are enabled.
+     *
+     * @type sai_s32_list_t sai_port_interface_type_t
+     * @flags CREATE_AND_SET
+     * @default empty
+     */
+    SAI_PORT_ATTR_ADVERTISED_INTERFACE_TYPE,
+```
+
+Please note that `SAI_PORT_ATTR_ADVERTISED_INTERFACE_TYPE` is a new attribute introduced in SAI 1.7.1. Vendors need to implement this attribute in their SAI implementation.
+
+### Configuration and management 
+
+#### CLI Enhancements
+
+A few new CLI commands are designed to support port auto negotiation.
+
+##### Config auto negotiation mode
+
+```
+Format:
+  config interface autoneg <interface_name> <mode>
+
+Arguments:
+  interface_name: name of the interface to be configured. e.g: Ethernet0
+  mode: auto negotiation mode, can be either "enabled" or "disabled"
+
+Example:
+  config interface autoneg Ethernet0 enabled
+  config interface autoneg Ethernet0 disabled
+
+Return:
+  error message if interface_name or mode is invalid otherwise empty
+```
+
+##### Config advertised speeds
+
+Configuring advertised speeds takes effect only if auto negotiation is enabled. If auto negotiation is disabled, this command still saves advertised speeds value to CONFIG_DB.
+
+```
+Format:
+  config interface advertised-speeds <interface_name> <speed_list>
+
+Arguments:
+  interface_name: name of the interface to be configured. e.g: Ethernet0
+  speed_list: a list of speeds to be advertised or "all". e.g: 40000,100000.
+
+Example:
+  config interface advertised-speeds Ethernet0 40000,100000
+  config interface advertised-speeds Ethernet0 all
+
+Return:
+  error message if interface_name or speed_list is invalid otherwise empty
+
+Note:
+  speed_list value "all" means all supported speeds 
+```
+
+This command always replace the advertised speeds instead of append. For example, say the current advertised speeds value are "10000,25000", if user configure it with `config interface advertised-speeds Ethernet0 40000,100000`, the advertised speeds value will be changed to "40000,100000".
+
+##### Config interface type
```

Comment:
The interface type should be determined by the media that is plugged in (e.g., copper DAC, optical transceiver, etc.). Hence SONiC (i.e., `xcvrd`) should be able to dynamically set the interface type based on the detected media using the `media_settings.json` framework. That framework today is used to set pre-emphasis values, but it can/should be extended to perform other default settings such as interface type, FEC, auto-negotiation, etc.

This config could override what is defined in the platform's `media_settings.json` file, but care should be taken when the configured interface type differs from what is actually supported by the media (e.g., setting `CR` (copper type) for an optical transceiver).

Reply:
Hi @jeff-yin, thanks for the comments. As you mentioned, media_settings.json is used to set the pre-emphasis values for ports. And it handles it this way:

1. When xcvrd daemon start, it reads the media_setting.json and set pre-emphasis values to APPL_DB for each port
2. When a new cable is inserted, xcvrd will use the value in media_setting.json to set pre-emphasis value to APPL_DB for this port.

As xcvrd just reads the configuration from media_setting.json and set value to APPL_DB, xcvrd does not care what the configuration is. So if you want to use media_setting.json to specify the default value for autoneg, adv_speeds, fec and so on, nothing needs to be changed. However, there is something that user need to understand:

**_If we have autoneg, adv_speeds or other port configuration both in CONFIG_DB and media_settings.json, the value in media_settings.json will be used if user reboot switch, restart pmon docker or remove/insert the cable._**

I suppose it means that, if user choose to configure port autoneg in media_settings.json, he/she should not configure port autoneg via CLI, or what he/she configured in CLI will lose effect after rebooting/restart pmon/re-insert SFP.

Reply:
Hi @jeff-yin, the problem i see with setting default based on standards in media_settings.json is it would break the backward compatibility. Currently SONiC is deployed with autoneg and media type not being set. Setting this automatically via media_settings.json would cause interop issues and in many cases the link will not even come up. Hence currently we can go with user configuration approach alone and later have discussion on how to implement based on standards